src/bcrypt_nif_worker.erl

```diff
@@ -14,39 +14,72 @@
          handle_call/3, handle_cast/2, handle_info/2]).
 
 -record(state, {
-          default_log_rounds,
-          context
+          current_ctx_index :: pos_integer(),
+          contexts :: tuple()
          }).
 
-start_link() -> gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+-define(DEFAULT_LOG_ROUNDS, 12).
 
-gen_salt() -> gen_server:call(?MODULE, gen_salt, infinity).
-gen_salt(Rounds) ->
-    gen_server:call(?MODULE, {gen_salt, Rounds}, infinity).
-hashpw(Password, Salt) ->
-    gen_server:call(?MODULE, {hashpw, Password, Salt}, infinity).
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+start_link() ->
+    DefultBCryptQueueNum = erlang:system_info(schedulers),
+    start_link(DefultBCryptQueueNum).
 
-init([]) ->
-    {ok, Default} = application:get_env(bcrypt, default_log_rounds),
-    Ctx = bcrypt_nif:create_ctx(),
-    {ok, #state{default_log_rounds = Default, context = Ctx}}.
+start_link(BCryptQueueNum) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [BCryptQueueNum], []).
 
-terminate(shutdown, _) -> ok.
+gen_salt() ->
+    Rounds = application:get_env(bcrypt, default_log_rounds, ?DEFAULT_LOG_ROUNDS),
+    gen_salt(Rounds).
+gen_salt(Rounds) ->
+    {ok, bcrypt_nif:gen_salt(Rounds)}.
 
-handle_call(gen_salt, _From, #state{default_log_rounds = R} = State) ->
-    {reply, {ok, bcrypt_nif:gen_salt(R)}, State};
-handle_call({gen_salt, R}, _From, State) ->
-    {reply, {ok, bcrypt_nif:gen_salt(R)}, State};
-handle_call({hashpw, Password, Salt}, _From, #state{context=Ctx}=State) ->
+hashpw(Password, Salt) ->
     Ref = make_ref(),
```

Comment:
Nit: we don't need this one since we now have `MRef` from `monitor`.

Reply:
Would be nice use alias to squash pid & ref and simplify the nif part a bit, but seems there are still no nifs for sending messages to alias refs 😓